### PR TITLE
(#3) Enabler: Create Search Box

### LINKED ISSUE #3: Enabler: Create Search Box
# Create the Search Box

This should be pretty much a lift of the search box (Autocomplete + A-Z). 

The Autocomplete may have a differently structured endpoint if using the legacy API. The new API should be the same structure.

### Acceptance Criteria
Yes, the search box does not live on its own, but wherever it shows the following should hold true... These will be written from the context of the term definition page.
```gherkin
Scenario: As a user, I would like to see an option to search when I am viewing the definition for a term
Given the user is viewing the definition with the pretty url "napabucasin"
  Then heading "Search NCI's Drug Dictionary" appears
    Then search options for "Starts with" and "Contains" appears
    And "Starts with" radio is selected by default
    And "keywords" search box appears
    And search button appears beside search box with "Search"
    And "Browse:" appears with A-Z List of Links beside it
    And each option appears as a link
```

```gherkin
Scenario: User is able to see autosuggested items when typing 3 or more characters into the search box and has selected "Starts With"
Given the user is viewing the definition with the pretty url "napabucasin"
 And "Starts with" radio is selected by default
    And search bar contains a placeholder text "Enter keywords or phrases"
    When user clicks on the search bar
    Then helper text "Please enter 3 or more characters" appears
    When user types "/cd" in the search bar
    Then placeholder text disappears
    And Autosuggest appears after user types in 3 or more characters
    Then highlighting of the text "/cd" appears in the autosuggest field
```

```gherkin
Scenario: User is able to see autosuggested items when typing 3 or more characters into the search box and has selected the Contains option
Given the user is viewing the definition with the pretty url "napabucasin"
When user selects "Contains" option
    Then search bar contains a placeholder text "Enter keywords or phrases"
    When user types "bin" in the search bar
    Then placeholder text disappears
    And Autosuggest appears after user types in 3 or more characters
    Then highlighting of the text "bin" appears in the autosuggest field
    When user selects first autosuggested term
    Then term is populated into the search bar
```

**Analytics**
```gherkin
  Scenario: Analytics click event on A-Z list is raised when user clicks on a letter in the list.
    Given "dictionaryTitle" is set to "NCI Drug Dictionary"
       And "baseHost" is set to "http://localhost:3000"
       And "canonicalHost" is set to "https://www.cancer.gov"
       And "siteName" is set to "National Cancer Institute"
       And "channel" is set to "Publications"
       And "analyticsPublishedDate" is set to "02/02/2011"
       And "analyticsName" is set to "Drug"
    When the user is viewing the definition with the pretty url "bevacizumab"
    And the page title is "bevacizumab"
    # And browser waits
    Then the NCIDataLayer is cleared
    When user selects letter "A" from A-Z list
    Then search results page displays results title "# results found for: A"
    Then there should be an analytics event with the following details
      | key                    | value                          |
      | type                   | Other                          |
      | event                  | DrugDictionaryApp:Other:AZClick      |
      | linkName               | DrugDictionarySearchAlphaList |
      | data.dictionaryTitle   | NCI Drug Dictionary |
      | data.analyticsName     | Drug                    |
      | data.letter						 | A                              |
```

```gherkin
	Scenario: When a user clicks on the search button for a starts with search, an analytics click event will be fired.
    Given "dictionaryTitle" is set to "NCI Drug Dictionary"
       And "baseHost" is set to "http://localhost:3000"
       And "canonicalHost" is set to "https://www.cancer.gov"
       And "siteName" is set to "National Cancer Institute"
       And "channel" is set to "Publications"
       And "analyticsPublishedDate" is set to "02/02/2011"
       And "analyticsName" is set to "Drug"
    When the user is viewing the definition with the pretty url "bevacizumab"
       And the page title is "bevacizumab"
      And the search box has loaded
      And the user selects the "Starts with" option in the search box
      And the user enters "beva"
      And the user submits the keyword search
    Then there should be an analytics event with the following details
			| key                  | value                           |
			| type                 | Other                           |
			| event                | DrugDictionaryApp:Other:KeywordSearch |
			| linkName             | DrugDictionarySearch           |
			| data.dictionaryTitle | NCI Drug Dictionary  |
			| data.analyticsName   | Drug                     |
			| data.searchTerm      | beva                      |
			| data.searchType      | starts with                     |
```

```gherkin
	Scenario: When a user clicks on the search button for a contains search, an analytics click event will be fired.
    Given "dictionaryTitle" is set to "NCI Drug Dictionary"
       And "baseHost" is set to "http://localhost:3000"
       And "canonicalHost" is set to "https://www.cancer.gov"
       And "siteName" is set to "National Cancer Institute"
       And "channel" is set to "Publications"
       And "analyticsPublishedDate" is set to "02/02/2011"
       And "analyticsName" is set to "Drug"
    When the user is viewing the definition with the pretty url "bevacizumab"
       And the page title is "bevacizumab"
       And the search box has loaded
       And the user selects the "Contains" option in the search box
       And the user enters "beva"
       And the user submits the keyword search
    Then there should be an analytics event with the following details
			| key                  | value                           |
			| type                 | Other                           |
			| event                | DrugDictionaryApp:Other:KeywordSearch |
			| linkName             | DrugDictionarySearch           |
			| data.dictionaryTitle | NCI Drug Dictionary  |
			| data.analyticsName   | Drug                     |
			| data.searchTerm      | beva                      |
			| data.searchType      | contains                        |
```

> **ESTIMATE** With Tests: 5, Without Tests: 3

### Resources:
* Feature: JIRA Ticket #
* Mockups: *Here goes a URL to or the name of the mockup(s) in inVision*
* References: *URL to web site with examples*
* Web Page: *Here goes a URL to an existing web page*

### Prerequisites
- *required dependency for identifying the problem, scope and then solution for this issue*
- [ ] Short Spike Description - Issue #9999

### Sub-Tasks
- [ ] Fetch autosuggest
- [ ] Build the search box
- [ ] Update AC for SearchBox
- [ ] Integration tests

### Notes
* This will not display results, result display is a separate story. This is just for the Autocomplete box, Starts With/Contains radio buttons and the A-Z list.

#### Autosuggest API URLs:
* **begins** `https://www.cancer.gov/Dictionary.Service/v1/searchSuggest?dictionary=drug&language=English&searchType=begins&searchText=SEARCH_TEXT`
* **contains** `https://www.cancer.gov/Dictionary.Service/v1/searchSuggest?dictionary=drug&language=English&searchType=contains&searchText=SEARCH_TEXT` 

#### The logic on prod for AUTOSUGGEST
Starts with:
* Starts with DrugTerm
* Starts with DrugAlias
* Contains Drug Alias

Contains:
* Contains Drug Term

#### Technical implementation for Starts with:
API needs to make three separate queries. 

1. Starts with DrugTerm
2. Starts with DrugAlias
3. Contains Drug Alias

If first query returns >= 10 results, the second and third queries are not needed. If first query returns < 10 results, the second query's result is appended after first query's result (the result of the third query is discarded if the merged results of the first and second queries is >= 10). Results of the third query is added if 1st and 2nd merged results is < 10.

Contains: is just a pure contains

### PR DESCRIPTION
**Pull request by seyilonge-nci**
Closes #3

* Added SearchBox component
* Updated unit tests
* Updated dependent components

### CODE REVIEW

File: src/components/atomic/az-list/AZList.jsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+		});
+	};
+
+	return (
+		<nav className="az-list" data-testid="tid-az-list">
+			<span className="browse">{azListLabel}:</span>
+			<ul>
+				{AZListArray.map((item, i) => {
+					const expandChar =
+						item === '#' ? '%23' : item !== 'All' ? item.toUpperCase() : item;
```
Comment (arcepaul) [inline src/components/atomic/az-list/AZList.jsx:32]: nit: please reformat for readability

File: src/components/molecules/search/Search.jsx [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+			selectedOption &&
+			getKeyValueFromObject(selectedOption, searchMatchTypeAnalyticsMap)
+				? getKeyValueFromObject(selectedOption, searchMatchTypeAnalyticsMap)
+				: searchMatchTypeAnalyticsMap[searchMatchType.beginsWith];
+		tracking.trackEvent({
+			type: 'Other',
+			event: 'DrugDictionaryApp:Other:KeywordSearch',
+			linkName: 'DrugDictionarySearch',
+			searchTerm: searchText,
+			searchType: searchType,
```
Comment (KateMashkinaNIH) [inline src/components/molecules/search/Search.jsx:64]: The event is missing one parameter :  | analyticsName     | Drug |

File: src/components/atomic/az-list/AZList.jsx [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	useEffect(() => {
+		AZListArray.push('All');
+	}, []);
+
+	const handleAZLinkClick = (letter) => {
+		tracking.trackEvent({
+			type: 'Other',
+			event: 'DrugDictionaryApp:Other:AZClick',
+			linkName: 'DrugDictionarySearchAlphaList',
+			letter,
```
Comment (KateMashkinaNIH) [inline src/components/atomic/az-list/AZList.jsx:22]: The event should have one more param : |analyticsName     | Drug |

**Review by KateMashkinaNIH — CHANGES_REQUESTED** [review]

**Review by karlikpj — APPROVED** [review]

File: src/components/atomic/az-list/AZList.scss [hunk 4]
```diff
@@ -0,0 +1,15 @@
+.az-list {
+	.browse {
+		float: none;
+	}
+
```
Comment (arcepaul) [inline src/components/atomic/az-list/AZList.scss:5]: why all the `&`s?

File: src/views/Definition/Definition.jsx [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+import { SearchBox } from '../../components';
+
+const Definition = () => {
+	const [{ dictionaryTitle, dictionaryIntroText }] = useStateValue();
+
+	return (
+		<>
+			<h1>{dictionaryTitle}</h1>
+			<div>
```
Comment (arcepaul) [inline src/views/Definition/Definition.jsx:13]: is this wrapper necessary?

File: src/components/atomic/az-list/AZList.jsx (hunk 1 above)
Comment (seyilonge-nci) [reply src/components/atomic/az-list/AZList.jsx:32]: Done!

File: src/components/molecules/search/Search.jsx (hunk 2 above)
Comment (seyilonge-nci) [reply src/components/molecules/search/Search.jsx:64]: Added

File: src/components/atomic/az-list/AZList.jsx (hunk 3 above)
Comment (seyilonge-nci) [reply src/components/atomic/az-list/AZList.jsx:22]: Added

File: src/components/atomic/az-list/AZList.scss (hunk 4 above)
Comment (seyilonge-nci) [reply src/components/atomic/az-list/AZList.scss:5]: Removed

File: src/views/Definition/Definition.jsx (hunk 5 above)
Comment (seyilonge-nci) [reply src/views/Definition/Definition.jsx:13]: No it's not. Removed.

**Review by kate-mashkina — APPROVED** [review]

**Review by arcepaul — APPROVED** [review]

File: src/components/molecules/search/Search.jsx (hunk 2 above)
Comment (bryanpizzillo) [reply src/components/molecules/search/Search.jsx:66]: Don't most of the click events and page loads have analyticsName? If so it should be wired up at the analytics provider level. See https://github.com/NCIOCPL/glossary-app/blob/develop/src/index.js#L85

File: src/views/Definition/Definition.jsx [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+const Definition = () => {
+	const [{ dictionaryTitle, dictionaryIntroText }] = useStateValue();
+
+	return (
+		<>
+			<h1>{dictionaryTitle}</h1>
+			<p
+				dangerouslySetInnerHTML={{
+					__html: dictionaryIntroText,
```
Comment (bryanpizzillo) [inline src/views/Definition/Definition.jsx:15]: What is this doing here? There should not be any intro text except the home page. Can you make this like
```
<h1>Definition Title Here</h1>
<p>Definition Here</p>
```
Or really, just have nothing except your search box.

**Review by bryanpizzillo — CHANGES_REQUESTED** [review]

File: src/views/Definition/Definition.jsx (hunk 6 above)
Comment (seyilonge-nci) [reply src/views/Definition/Definition.jsx]: Arghhh! Quite true, no intro text. Removed and put in placeholders for definition title and definition.

Comment (seyilonge-nci) [reply src/views/Definition/Definition.jsx]: Moved analyticsName to provider.

**Review by arcepaul — APPROVED** [review]

**Review by KateMashkinaNIH — APPROVED** [review]

**Pull request merged**